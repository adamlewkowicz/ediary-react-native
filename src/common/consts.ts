export const PRODUCT_UNITS = ['g', 'mg', 'kg', 'ml', 'l'] as const;
export const BASE_MACRO_ELEMENTS = ['carbs', 'prots', 'fats'] as const
export const MACRO_ELEMENTS = [...BASE_MACRO_ELEMENTS, 'kcal'] as const;