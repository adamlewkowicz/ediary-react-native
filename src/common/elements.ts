
export const elements = {
  Home: {
    label: {
      toggleMealButton: 'Pokaż szczegóły posiłku',
      addMealProductNavButton: 'Wyszukaj produkt do posiłku',
      productQuantityInput: 'Zmień ilość produktu'
    },
  },
  ProductFind: {
    placeholder: {
      searcherInput: 'Nazwa produktu',
      foundProductButton: 'Dodaj produkt do posiłku'
    }
  }
}