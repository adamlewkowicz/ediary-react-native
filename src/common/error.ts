export class UserIdCannotBeNullError extends Error {}
export class UnsupportedUnitTypeError extends Error {}
export class TransactionManagerUndefinedError extends Error {
  constructor(message = 'Transaction manager is unavailable') {
    super();
  }
}