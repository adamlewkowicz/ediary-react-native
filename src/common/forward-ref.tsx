import React, { RefObject } from 'react';

interface ComponentProps {
  forwardedRef: RefObject<unknown>
  ref: RefObject<unknown>
}

export function forwadRef<T>(Component: T): T {
  return React.forwardRef(
    (props, ref) => <Component {...props} forwadedRef={ref as any} />
  ) as any;
}