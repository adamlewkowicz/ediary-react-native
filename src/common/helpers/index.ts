export * from './measure';
export * from './values';