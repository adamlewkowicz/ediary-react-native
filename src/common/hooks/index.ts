export * from './useUserId';
export * from './useConnected';