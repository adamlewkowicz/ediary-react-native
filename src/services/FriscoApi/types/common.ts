export interface FriscoProductId extends String {}