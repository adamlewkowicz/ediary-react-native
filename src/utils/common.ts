import { ToastAndroid, Alert, LayoutAnimation } from 'react-native';
import { Client } from 'bugsnag-react-native';
import { BUGSNAG_API_KEY } from 'react-native-dotenv';
import { IS_DEV } from '../common/consts';

const bugsnag = new Client(BUGSNAG_API_KEY);

export const createDebouncedFunc = <T extends (...args: any) => any>(
  callback: T,
  delay = 250
) => {
  let timeout: NodeJS.Timeout;

  return (...args: Parameters<T>) => {
    clearTimeout(timeout);
    timeout = setTimeout(() => callback(args), delay);
  }
}

export const toastCenter = (message: string): void => {
  ToastAndroid.showWithGravity(
    message,
    ToastAndroid.SHORT,
    ToastAndroid.CENTER
  );
}

export const alertDelete = (
  title: string,
  message: string,
  onDeleteConfirmed: () => void  
): void => {
  Alert.alert(
    title,
    message,
    [
      {
        text: 'Anuluj',
        style: 'cancel'
      },
      {
        text: 'UsuÅ„',
        onPress: onDeleteConfirmed
      }
    ]
  );
}

export const layoutAnimateEase = (onAnimationDidEnd?: () => void) => {
  LayoutAnimation.configureNext(
    LayoutAnimation.Presets.easeInEaseOut,
    onAnimationDidEnd
  );
}

export const toLocaleString = (value: number) => new Intl.NumberFormat('pl-PL').format(value);

export const handleError = (error: Error): void => {
  if (IS_DEV) {
    throw error;
  } else {
    bugsnag.notify(error);
  }
}