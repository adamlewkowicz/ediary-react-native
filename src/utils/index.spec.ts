import {
  mapAsyncSequence,
  filterByUniqueId,
  findOrFail,
  sortByMostAccurateName,
} from '.';
import { createDebouncedFunc } from './common';

test('createDebouncedFunc - debounced function gets called only once', () => {
  jest.useFakeTimers();
  const actionMock = jest.fn();

  const debounceTime = 150;
  const debouncedAction = createDebouncedFunc(actionMock, debounceTime);

  debouncedAction(actionMock, debounceTime);
  debouncedAction(actionMock, debounceTime);

  jest.advanceTimersByTime(debounceTime);

  expect(actionMock).toHaveBeenCalledTimes(1);
});

test('mapAsyncSequence - runs promises in a sequence', async () => {
  const promiseMock = jest.fn().mockImplementation(async () => {});
  const dataMock = [1, 2];

  await mapAsyncSequence(dataMock, value => promiseMock(value));

  expect(promiseMock).toHaveBeenCalledTimes(dataMock.length);
  expect(promiseMock).toHaveBeenNthCalledWith(1, dataMock[0]);
  expect(promiseMock).toHaveBeenNthCalledWith(2, dataMock[1]);
});

test('filterByUniqueId - filters items by unique id', () => {
  const items = [{ id: 1 }, { id: 1 }];
  const filteredItems = items.filter(filterByUniqueId);

  expect(filteredItems.length).toBeLessThan(items.length);
});

test('findOrFail - throws error if no item was found', () => {
  const items = [{ name: 'A' }];
  const subject = () => findOrFail(items, item => item.name === 'B');
  expect(subject).toThrowError();
});

test('sortByMostAccurateName - sorts items by most matching name', () => {
  const name = 'orange';
  const dataMock = [
    { name: 'Orange a' },
    { name: 'Eng' },
    { name: 'Orange' },
  ];

  const result = dataMock.sort(sortByMostAccurateName(name));

  expect(result).toMatchSnapshot();
});