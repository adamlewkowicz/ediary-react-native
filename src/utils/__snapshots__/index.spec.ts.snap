// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Utils sortByMostAccurateName should sort items by most matching name 1`] = `
Array [
  Object {
    "name": "Orange",
  },
  Object {
    "name": "Orange a",
  },
  Object {
    "name": "Eng",
  },
]
`;
